docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+            - `pin`: The user's entered pin code
+        - if the FIDO token is removed before the user is done interacting with the UI, the frontend emits a `passkeyRemoved` event
+        - Once the user touches the FIDO token, the authenticator will continue the converstaion on its own
+
+- `monitorSmartcard`: Prompt for the user to insert a smartcard
+    - The frontend monitors for smartcard insertion events, and prompts for the user to insert a smartcard
+    - If a smarcard is already inserted, the frontend responds immediately
+    - This method is either primary or secondary
+    - A frontend shouldn't advertise support for this mechanism in `supportedMechanisms`; support is implied for any frontend that supports the `smartcard` mechanism.
+    - Arguments:
+        - `primary`: When true, this is a primary authentication mechanism
+    - Response: frontend emits the `smartcardInserted` event when a smartcard is inserted
+
+- `smartcard`: Smart card handling
+    - Smartcards might have more than one certificate available. If this card has more than one certificate, the frontend first prompts the user to pick a certificate
+    - Next, the frontend prompts the user for a "pin" (which is actually an alphanumeric password)
+    - The frontend shall continue monitoring for smartcard insertion/removal events
+    - This method is always primary
+    - Arguments:
+        - `certificates`: An array of objects that describe the available certificates that the user can select between. Each object in the array has the following keys:
+            - `id`: The machine-usable string that identifies this certificate
+            - `slotName`: A display name for the PIV slot of this certificate. Example: "Certificate for PIV Authentication"
+            - `commonName`: The `CN` field of the certificate. Example: "John"
+            - `organization`: The `O` field of the certificate. Example: "Red Hat"
+            - TODO: Do we have the slot available in non-display-name-form? Are there other fields from the certificate that we should expose?
+    - Response:
+        - if the user has confirmed their entered pin code, then the frontend emits a `response` event with the following arguments:
+            - `pin`: The user's entered pin code
+            - `certificate`: The `id` of the certificate that the user selected. Required if `certificates` has more than one option
+        - if the smartcard is removed before the user is done interacting with the UI, the frontend emits a `smartcardRemoved` event
+
+## Other events
+
+While the primary conversation is occuring as described above, some additional events may occur. This section of the document explains these different events.
+
+#### General authentication status
+
+These events might be emitted while any mechanism is active:
+
+- `authenticationFailed`: The primary authentication mechanism has failed
```

Comment:
Perhaps we should have some way to prompt the user to roll back to their old password?

![](https://cdsassets.apple.com/live/7WUAS350/images/ios/ios-26-iphone-16-pro-iphone-unavailable-enter-previous-passcode.png)